### ISSUE TITLE
test_asyncio fails always

### ISSUE THREAD

**Issue by 0785d778-18fb-4fc2-891c-4d0b7758f730** [python/cpython#74075]
BPO | [29889](https://bugs.python.org/issue29889)
--- | :---
Nosy | @vstinner, @ned-deily, @asvetlov, @zware, @1st1, @zhangyangyu, @torsava

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2017-12-20.20:51:31.338>
created_at = <Date 2017-03-23.21:30:44.217>
labels = ['expert-asyncio']
title = 'test_asyncio fails always'
updated_at = <Date 2017-12-20.20:51:31.336>
user = 'https://bugs.python.org/ThomasKnox'
```

bugs.python.org fields:
```python
activity = <Date 2017-12-20.20:51:31.336>
actor = 'asvetlov'
assignee = 'none'
closed = True
closed_date = <Date 2017-12-20.20:51:31.338>
closer = 'asvetlov'
components = ['asyncio']
creation = <Date 2017-03-23.21:30:44.217>
creator = 'Thomas Knox'
dependencies = []
files = []
hgrepos = []
issue_num = 29889
keywords = []
message_count = 8.0
messages = ['290059', '290060', '290066', '290067', '290068', '290075', '293335', '308812']
nosy_count = 8.0
nosy_names = ['vstinner', 'ned.deily', 'asvetlov', 'zach.ware', 'yselivanov', 'xiang.zhang', 'torsava', 'Thomas Knox']
pr_nums = []
priority = 'normal'
resolution = 'works for me'
stage = 'resolved'
status = 'closed'
superseder = None
type = None
url = 'https://bugs.python.org/issue29889'
versions = ['Python 3.6']
```

</p></details>

**Comment by 0785d778-18fb-4fc2-891c-4d0b7758f730** [comment]
Downloaded Python 3.6.1 source code onto CentOS 7.3 (64 bit), Fedora 25 (64 bit), Ubuntu 16.10 (64 bit) and Raspberry Pi 8.0 (32 bit).

Configured with
./configure --enable-shared --enable-optimizations --enable-loadable-sqlite-extensions --disable-ipv6 --with-system-expat --with-system-ffi --with-threads

On every platform, when running the profile generation code, test_asyncio fails with this error message:

0:06:45 [ 25/405] test_asyncio
Executing \<Task finished coro=\<CoroutineTests.test_async_def_wrapped.\<locals\>.start() done, defined at /home/pi/Source/Python-3.6.1/Lib/test/test_asyncio/test_pep492.py:150\> result=None created at /home/pi/Source/Python-3.6.1/Lib/asyncio/base_events.py:446\> took 2.106 seconds
0:13:11 [ 26/405] test_asyncore -- test_asyncio failed in 6 min 27 sec

**Comment by ned-deily** [comment]
There have been other reports (e.g. bpo-29712) of failures when using --enable-optimizations with --enable-shared.  Can you see if removing --enable-shared makes a difference for your configurations?

**Comment by 0785d778-18fb-4fc2-891c-4d0b7758f730** [comment]
I changed the configuration flags to be:

pi@pi3:~/Source/Python-3.6.1 $ ./configure --enable-optimizations --enable-loadable-sqlite-extensions --disable-ipv6 --with-system-expat --with-system-ffi --with-threads

The profiling filed again with:

0:04:47 [ 25/405] test_asyncio
Executing \<Task finished coro=\<CoroutineTests.test_async_def_wrapped.\<locals\>.start() done, defined at /home/pi/Source/Python-3.6.1/Lib/test/test_asyncio/test_pep492.py:150\> result=None created at /home/pi/Source/Python-3.6.1/Lib/asyncio/base_events.py:446\> took 1.187 seconds
0:08:45 [ 26/405] test_asyncore -- test_asyncio failed in 3 min 59 sec

But at least it failed 2.5 minutes faster. :)

**Comment by zware** [comment]
Does the build fail, or just the test in the profile generation?

**Comment by 0785d778-18fb-4fc2-891c-4d0b7758f730** [comment]
The build succeeds, but the profile generation fails (only the test_asyncio portion fails), then does a new profiled build after the profiling run completes which succeeds.

Then, when doing a 'make install', it builds the entire thing again and runs another profile generation run which also fails, then a new profiled build. Possibly because the first profile run failed?

On an RPi, this is exceedingly painful as a build/profile/build cycle takes about 10 hours. So doing it twice is double-unfun.

**Comment by zhangyangyu** [comment]
> Then, when doing a 'make install', it builds the entire thing again and runs another profile generation run which also fails, then a new profiled build. Possibly because the first profile run failed?

This is a known issue. Some build commands always clear the environment and rebuild. See bpo-29243.

**Comment by vstinner** [comment]
Can someone test again with a development branch? The bpo-29243 has been fixed in 2.7, 3.5, 3.6 and master (3.7) branches thanks to Tomas Orsava!
